RuntimeException: Cannot calculate hash for empty file

It would be good to add a file name to the exception message to make getting know what caused that problem easier.

```
Caused by: java.lang.RuntimeException: Cannot calculate hash for empty file.
        at com.matthewtamlin.retrial.hash.Sha512HashGenerator$generateHash$1.call(Sha512HashGenerator.kt:36)
        at com.matthewtamlin.retrial.hash.Sha512HashGenerator$generateHash$1.call(Sha512HashGenerator.kt:28)
        at io.reactivex.internal.operators.single.SingleFromCallable.subscribeActual(SingleFromCallable.java:44)
```

Code to reproduce that issue: https://github.com/szpak/gradle-pitest-plugin/tree/tech/retrial

Please note that you need to add `retrail.jar` on your own. Pushing to the remote repository I noticed that it takes 50M+. I have to consider if it is not too much for my git repo.

As a side note I was wondering what do you do with the dependencies having JAR at 50KB of size (I noticed KB -> MB on long lasting push :) ). As you use Kotlin and RxJava it may be hard. Maybe at least Kotlin could be taken from the Gradle dependencies itself (if provided as of 4.x)?

--- a/src/main/kotlin/com/matthewtamlin/retrial/hash/Sha512HashGenerator.kt
+++ b/src/main/kotlin/com/matthewtamlin/retrial/hash/Sha512HashGenerator.kt
@@ -26,17 +26,20 @@ import javax.inject.Provider
  * Generates SHA2-512 hashes from files.
  */
 class Sha512HashGenerator {
+
   private val sha512DigestProvider: Provider<MessageDigest> = Provider { MessageDigest.getInstance("SHA-512") }
 
   /**
    * Generates a SHA2-512 hash of the contents of a file.
    */
   fun generateHash(file: File) = Single.fromCallable {
-    if (file.length() == 0L) {
-      throw RuntimeException("Cannot calculate hash for empty file.")
-    }
+    if (file.length() == 0L) hashForEmptyString() else hashForFile(file)
+  }
 
-    sha512DigestProvider
+  private fun hashForEmptyString() = Sha512Hash("E3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B934CA495991B7852B855")
+
+  private fun hashForFile(file: File): Sha512Hash {
+    return sha512DigestProvider
         .get()
         .digest(Files.readAllBytes(file.toPath()))
         .map { toHexString(it) }

--- a/src/test/kotlin/com/matthewtamlin/retrial/hash/Sha512HashGeneratorTests.kt
+++ b/src/test/kotlin/com/matthewtamlin/retrial/hash/Sha512HashGeneratorTests.kt
@@ -45,12 +45,12 @@ class Sha512HashGeneratorTests {
   inner class TestGenerateHash {
 
     @Test
-    @DisplayName("should fail when the file is empty")
+    @DisplayName("should return the SHA2-512 hash of an empty string when the file is empty")
     fun fileIsEmpty() {
       generator
           .generateHash(file)
           .test()
-          .assertError(RuntimeException::class.java)
+          .assertResult(Sha512Hash("E3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B934CA495991B7852B855"))
     }
 
     @Test

